pennylane/operation.py

```diff
@@ -528,6 +529,123 @@ def __init__(self, *args, wires=None, do_queue=True):
+
+    The ``@`` symbol can be used as a tensor product operation:
+
+    >>> T = qml.PauliX(0) @ qml.Hadamard(2)
+    """
+    return_type = None
+    tensor = True
+    par_domain = None
+
+    def __init__(self, *args): #pylint: disable=super-init-not-called
+        self.obs = []
+
+        for o in args:
+            if isinstance(o, Tensor):
+                self.obs.extend(o.obs)
+            elif isinstance(o, Observable):
+                self.obs.append(o)
+            else:
+                raise ValueError("Can only perform tensor products between observables.")
+
+    def __str__(self):
+        """Print the tensor product and some information."""
+        return 'Tensor product {}: {} params, wires {}'.format([i.name for i in self.obs], len(self.params), self.wires)
+
+    @property
+    def name(self):
+        """All consituent observable names making up the tensor product.
+
+        Returns:
+            list[str]: list containing all observable names
+        """
+        return [o.name for o in self.obs]
+
+    @property
+    def num_wires(self):
+        """Number of wires the tensor product acts on.
+
+        Returns:
+            int: number of wires
+        """
+        return len([w for sublist in [o.wires for o in self.obs] for w in sublist])
+
+    @property
+    def wires(self):
+        """All wires in the system the tensor product acts on.
+
+        Returns:
+            list[list[Any]]: nested list containing the wires per observable
+                in the tensor product
+        """
+        return [o.wires for o in self.obs]
+
+    @property
+    def params(self):
+        """Raw parameters of all constituent observables in the tensor product.
+
+        Returns:
+            list[Any]: flattened list containing all dependent parameters
+        """
+        return [p for sublist in [o.params for o in self.obs] for p in sublist]
+
+    @property
+    def num_params(self):
+        """Raw parameters of all constituent observables in the tensor product.
+
+        Returns:
+            list[Any]: flattened list containing all dependent parameters
+        """
+        return len(self.params)
+
+    @property
+    def parameters(self):
+        """Evaluated parameter values of all constituent observables in the tensor product.
+
+        Returns:
+            list[list[Any]]: nested list containing the parameters per observable
+                in the tensor product
+        """
+        return [o.parameters for o in self.obs]
+
+    def __matmul__(self, other):
+        if isinstance(other, Tensor):
+            self.obs.extend(other.obs)
+            return self
+
+        if isinstance(other, Observable):
+            self.obs.append(other)
+            return self
+
+        raise ValueError("Can only perform tensor products between observables.")
+
+    def __rmatmul__(self, other):
+        if isinstance(other, Observable):
+            self.obs[:0] = [other]
+            return self
```

Comment:
I guess ordering ambiguities (`X @ Y` vs `Y @ X`) are resolved by the `wires` argument, so this is just for aesthetics?

Reply:
Yep. `__matmul__` takes into account left multiplication of a observable with something, this takes into account right multiplication. It's only an issue since `expval @ tensor` is possible.